### ISSUE TITLE
Decide and implement sentence text objects (is and as)

### ISSUE THREAD

**Issue by lervag** [lervag/vimtex#437]
After #320, it seems relevant to define a set of rules for the sentence text objects. Sentences should include inline maths, at least in the `$...$` variant. It should also respect environment boundaries, even though this might not always be semantically correct.

I will update this post based on a (possibly) following discussion. I propose that we define a set of contexts and define how the text object should work in these contexts. This will make a good basis for a test driven development of the sentence text objects.

**Comment by lervag** [comment]
I'm curious: Is this something people are still interested in? @nilsbecker @Rmano

**Comment by nilsbecker** [comment]
i would still be interested in it in principle. but it's not clear if it's worth the effort now. my original stumbling block was that `s` was not nested inside `p`. is that still the case? or have you reverted to plain-vim `p` as text between empty lines? in the latter case, plain vim `s` and `p` would be a good solution as well, because they are easy to remember.

**Comment by Rmano** [comment]
(Sorry, a bit disconnected, summer time). I have still to check the work on `gq` --- but yes, I am quite in agreement with @nilsbecker  --- provided that inline math is "transparent" to `s`. But I also I am not sure if the effort is worth the while....

**Comment by lervag** [comment]
@nilsbecker I've reverted to plain-vim `p`.

@Rmano I don't quite understand what you mean with "provided that inline math is "transparent" to `s`. When I test it, I find that `vis` will selected the entire following line:

```tex
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, $f(x) = 2$ sed diam
nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
voluptua.
```

If I understand you guys correctly, then it seems that the plain-vim version of `s` is acceptable for now. Personally, I don't use the `s` text object very often, and so for me it is not an important feature, but if anyone should think otherwise, then I will consider to add it. If so, please try to define explicitly how it should work. That is, which boundaries should be considered a sentence boundary (both for the `is` and the `as` versions).

**Comment by Rmano** [comment]
I too use sparely it, so it's not so important, but... for completness:

```
\documentclass[border=5mm,tikz]{standalone}
\usepackage{tikz}
\usetikzlibrary{matrix}

\begin{document}
  \begin{tikzpicture}
    \draw[ultra thick,blue,->] (M-3-2.west) -- (M-2-2.west);
  \end{tikzpicture}
  
Lorem ipsum dolor POS1 sit amet, consetetur sadipscing elitr, $f(x) = 2$ sed diam
nonumy eirmod tempor invidunt 
ut labore et dolore magna aliquyam erat, sed diam
voluptua. Lorem ipsum dolor sit amet, consetetur 
sadipscing POS2 elitr, $f(x) = 2. $
sed diam nonumy eirmod tempor  invidunt  ut labore et 
dolore magna aliquyam erat, sed diam
voluptua. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, $f(x) = 2$ sed diam
nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
voluptua.

\end{document}
```

If I put the cursor on `POS1`, the command `vis` will select all the previous environment too, even if there is a blank line (a paragraph separator) there. I think that the "sentence" should finish whenever a paragraph separator is found. 
If I go to `POS2`, the same `vis` will break the math formula --- I know it's a bit contrived, but this is what I meant with "transparent" --- a `. ` in a math formula should not trigger the end of the sentence.

Again, I think this is a very low priority thing...

**Comment by lervag** [comment]
If I do `vim -Nu NONE test.tex`, where `test.tex` is your example file, then `vis` works as expected for POS1. For POS2, I am not sure I agree that it should let math mode be transparent. I think you should move the dot after the ending `$`.

For POS1, I suspect you are using options and/or plugins that alter the behaviour of `is`.

**Comment by Rmano** [comment]
Hmm... tried with the `vim -Nu none test.tex` but I still have the "tikzpicture" environment select by a `vis` on POS1. So it's probably version-dependent... I have mine there:

```
 [romano:~/tmp] 1 % vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Jun 16 2016 10:50:38)
Included patches: 1-1689
Modified by pkg-vim-maintainers@lists.alioth.debian.org
Compiled by pkg-vim-maintainers@lists.alioth.debian.org
Huge version with GTK2-GNOME GUI.  Features included (+) or not (-):
[...]
```

**Comment by lervag** [comment]
Ah, interesting. I'm on Vim 7.4.1910 on Arch Linux. I guess there's been a patch that fixes this, then. I'm compiling Vim to bisect it now.

**Comment by lervag** [comment]
Strange. I just compiled Vim Huge version 7.4.1689, and I still can't reproduce. I compiled in `~/vim`, so the executable becomes `~/vim/src/vim`. I do `~/vim/src/vim -Nu NONE test.tex`, where `test.tex` is your test file (unmodified). Then I go to `POS1` and I do `vis`. It now selects, as expected, from the beginning of the sentence to the first `.`.

**Comment by danielbmarques** [comment]
I don't know if it's relevant, but I'm using MacVim 7.4.1831 and it works properly here too.

**Comment by lervag** [comment]
Ah. IMHO, you should try to avoid trailing spaces, as they can have unexpected effects in other places as well. In this particular case, I think we should conclude that the plain-vim version of `is`/`as` is sufficient.
